oodeel/methods/rmds.py

```diff
@@ -0,0 +1,123 @@
+        super().__init__(eps=eps)
+
+    def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
+        """
+        Constructs the pear class means and the covariance matrice,
+        as well as the background mean and covariance matrice,
+        from ID data "fit_dataset".
+        The means and pseudo-inverses of the covariance matrices
+        will be used for RMDS score computation.
+
+        Args:
+            fit_dataset (Union[TensorType, DatasetType]): input dataset (ID)
+        """
+        # means and pseudo-inverse of the mean convariance matrice from Mahalanobis
+        super()._fit_to_dataset(fit_dataset)
+
+        # extract features
+        features, _ = self.feature_extractor.predict(fit_dataset)
+
+        # comput background mu and cov
+        _features_bg = self.op.flatten(features[0])
+        mu_bg = self.op.mean(_features_bg, dim=0)
+        _zero_f_bg = _features_bg - mu_bg
+        cov_bg = self.op.matmul(self.op.t(_zero_f_bg), _zero_f_bg) / _zero_f_bg.shape[0]
+
+        # background mu and pseudo-inverse of the mean covariance matrices
+        self._mu_bg = mu_bg
+        self._pinv_cov_bg = self.op.pinv(cov_bg)
+
+    def _score_tensor(self, inputs: TensorType) -> Tuple[np.ndarray]:
+        """
+        Computes an OOD score for input samples "inputs" based on the RMDS
+        distance with respect to the closest class-conditional Gaussian distribution,
+        and the background distribution.
+
+        Args:
+            inputs (TensorType): input samples
+
+        Returns:
+            Tuple[np.ndarray]: scores, logits
+        """
+        # input preprocessing (perturbation)
+        if self.eps > 0:
+            inputs_p = self._input_perturbation(inputs)
+        else:
+            inputs_p = inputs
+
+        # mahalanobis score on perturbed inputs
+        features_p, _ = self.feature_extractor.predict_tensor(inputs_p)
+        features_p = self.op.flatten(features_p[0])
+        gaussian_score_p = self._mahalanobis_score(features_p)
+
+        # background score on perturbed inputs
+        gaussian_score_bg = self._background_score(features_p)
+
+        # take the highest score for each sample
+        gaussian_score_corrected = self.op.max(
+            gaussian_score_bg - gaussian_score_p, dim=1
+        )
+        return -self.op.convert_to_numpy(gaussian_score_corrected)
+
+    def _background_score(self, out_features: TensorType) -> TensorType:
+        """
+        Mahalanobis distance-based background score. For each test sample, it computes
+        the log of the probability densities of some observations (assuming a
+        normal distribution) using the mahalanobis distance with respect to the
+        background distribution.
+
+        Args:
+            out_features (TensorType): test samples features
+
+        Returns:
+            TensorType: confidence scores (with respect to the background distribution)
+        """
+        mu = self._mu_bg
```

Comment:
Is this line necessary?

Reply:
This needs to change the following line as well. I have done it.